working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+including) that entry is an optional parameter with no default.
+
+For every entry with an initializer value, if no argument is passed for that
+parameter, the initializer value is assigned to that parameter in the
+initializer list of the constructor.  Note that this requires the ability to
+detect whether or not a parameter was passed, which is not expressible strictly
+as a de-sugaring.
+
+If no default constructor is defined in the class, a default constructor is
+generated which forwards to the generated primary constructor.
+
+**COMMENTARY(leafp):** *The treatment of initializers as default values here is
+  appealing, but it may be too cute.  It essentially adds non-const default
+  values only for this specific use case.  There's also a question of whether we
+  allow user defined constructors on the struct to also override the default
+  initializer, and if so via what syntax*.
+
+
+##### Alternative: named parameters
+
+These could be made named parameters.  This feels heavyweight, but has some
+advantages.  The restriction on names in the primary constructor would allow us
+to use the non-private versions of the field names as the parameter names.
+
+#### Equality
+
+If no equality method is defined in or inherited by the struct (except from
+Object), then an equality method will be defined which checks that its argument
+has the same runtime type as the receiver (actual runtime type, not the result
+of calling "runtimeType"), and that the fields of the two objects are pointwise
+equal.
+
+#### hashCode
+
+If no hashCode getter is defined in or inherited by the struct (except from
+Object), then a hashCode getter will be defined which hashes the runtime type
+(again, actual type) together with the hashes of each of the fields.
+
+#### toString
+
+If no toString method is defined in or inherited by the struct (except from
+Object), then a toString method will be defined which returns "struct".
+
+#### debugToString
+
+If no debugToString method is defined in or inherited by the struct, then a
+debugToString method will be defined which returns a formatted description of
+the receiver, of the form `<type>(<f0>, ..., <f1>)` where `<type>` is the
+runtime type of the receiver, and the `fi` are the result of calling
+"debugToString" on the `i`th field if that field is (dynamically) a struct, and
+otherwise the result of calling "toString".
+
+**COMMENTARY(leafp):** *This needs a bit of work.  If we keep the dynamic check
+  for "struct-ness", then we need do deal with the case that the user defines a
+  debutToString thing with the wrong type*.
+
+**COMMENTARY(leafp):** *If we keep this, we may wish to specify that compilers
+  may choose to make it a link time error to have an invocation of debugToString
+  in the program outside of asserts, and other debutToString methods*.
+
+
+#### copyWith
+
+If no copyWith method is defined in or inherited by the struct, then a copyWith
```

Comment:
What would an inherited `copyWith` method do that is 'correct'?

Reply:
Nothing useful.  Probably should just disallow it.

Reply:
I wonder whether this should just be a static method?  As long as we keep extension within the library, it works out, and it makes "extends" work out better.